src/audio_core/command_generator.cpp

```diff
@@ -0,0 +1,668 @@
+        // mix_info.
+        // TODO(ogniK): Depop to destination mix
+    } else if (in_params.splitter_info_id != AudioCommon::NO_SPLITTER) {
+        // TODO(ogniK): Depop to splitter
+    }
+
+    if (!depop) {
+        switch (in_params.sample_format) {
+        case SampleFormat::Pcm16:
+            DecodeFromWaveBuffers(voice_info, GetChannelMixBuffer(channel), dsp_state, channel,
+                                  worker_params.sample_rate, worker_params.sample_count,
+                                  in_params.node_id);
+            break;
+        case SampleFormat::Adpcm:
+            ASSERT(channel == 0 && in_params.channel_count == 1);
+            DecodeFromWaveBuffers(voice_info, GetChannelMixBuffer(0), dsp_state, 0,
+                                  worker_params.sample_rate, worker_params.sample_count,
+                                  in_params.node_id);
+            break;
+        default:
+            UNREACHABLE_MSG("Unimplemented sample format={}", in_params.sample_format);
+        }
+    }
+}
+
+void CommandGenerator::GenerateBiquadFilterCommandForVoice(ServerVoiceInfo& voice_info,
+                                                           VoiceState& dsp_state,
+                                                           s32 mix_buffer_count, s32 channel) {
+    for (std::size_t i = 0; i < AudioCommon::MAX_BIQUAD_FILTERS; i++) {
+        const auto& in_params = voice_info.GetInParams();
+        auto& biquad_filter = in_params.biquad_filter[i];
+        // Check if biquad filter is actually used
+        if (!biquad_filter.enabled) {
+            continue;
+        }
+
+        // Reinitialize our biquad filter state if it was enabled previously
+        if (!in_params.was_biquad_filter_enabled[i]) {
+            std::memset(dsp_state.biquad_filter_state.data(), 0,
+                        dsp_state.biquad_filter_state.size() * sizeof(s64));
+        }
+
+        // Generate biquad filter
+        GenerateBiquadFilterCommand(mix_buffer_count, biquad_filter, dsp_state.biquad_filter_state,
+                                    mix_buffer_count + channel, mix_buffer_count + channel,
+                                    worker_params.sample_count, voice_info.GetInParams().node_id);
+    }
+}
+
+void AudioCore::CommandGenerator::GenerateBiquadFilterCommand(
+    s32 mix_buffer, const BiquadFilterParameter& params, std::array<s64, 2>& state,
+    std::size_t input_offset, std::size_t output_offset, s32 sample_count, s32 node_id) {
+    if (dumping_frame) {
+        LOG_CRITICAL(Audio,
+                     "(DSP_TRACE) GenerateBiquadFilterCommand node_id={}, "
+                     "input_mix_buffer={}, output_mix_buffer={}",
+                     node_id, input_offset, output_offset);
+    }
+    const auto* input = GetMixBuffer(input_offset);
+    auto* output = GetMixBuffer(output_offset);
+
+    // Biquad filter parameters
+    const auto n0 = params.numerator[0];
```

Comment:
```suggestion
    const auto [n0, n1, n2] = params.numerator;
    const auto [d0, d1] = params.denominator;
    auto [s0, s1] = state;
```

Lets you concisely group them all together.